src/ot_orchestration/operators/gcs.py

```diff
@@ -51,3 +51,95 @@ def execute(self, context) -> None:
 
         blob = b.blob(self.path)
         blob.upload_from_filename(self.src)
+        self.log.info("uploaded file from %s to: %s", self.src, self.dst)
+
+
+class UploadRemoteFileOperator(BaseOperator):
+    """Custom operator that uploads a remote file to GCS.
+
+    This operator will create a GCS bucket if it does not exist and upload the
+    file from a URL to a path inside that bucket.
+
+    Args:
+        project_id: The GCP project ID. Defaults to the platform project.
+        src: The URL to the file to upload.
+        dst: The destination path in GCS.
+    """
+
+    template_fields: Sequence[str] = ("src", "dst")
+
+    def __init__(
+        self,
+        *args,
+        project_id: str = GCP_PROJECT_PLATFORM,
+        src: str,
+        dst: str,
+        **kwargs,
+    ) -> None:
+        super().__init__(*args, **kwargs)
+        self.project_id = project_id
+        self.dst = GCSPath(dst)
+        self.src = src
+
+        self.bucket_name, self.path = self.dst.split()
+
+    def execute(self, context) -> None:
+        """Execute the Operator."""
+        c = Client(project=self.project_id)
+        b = Bucket(client=c, name=self.bucket_name)
+        temp_file = Path("/tmp") / self.src.split("/")[-1]
```

Comment:
You could collect all chunks and save without the temporary file in this context.  Or even better fire the `UploadStringOperator` after reading the URI.

Reply:
But then if I download a 300GB file, I'd need a 300GB buffer. The chunking is there to lower memory requirements.

Regarding the second part, I think to use an operator inside another operator you have to subclass it. In this particular situation it increases the complexity and makes it way harder to follow (you have to go to the parent to check what's done), and the shared part is just four lines: create a bucket and upload the file. I'd argue it is better to repeat them in this case.

Reply:
You convinced be about skipping DRY.

If you keep 300GB payload, eventually you have to allocate as much fs memory. In that case it would be better to just transfer the chunk directly to the blob in one single loop. This way you do not need the memory for storing the temporary file as well.

Reply:
Indeed, but for that we need to write our own streaming method to upload to google cloud. Unless there is something in Google Cloud API to upload a stream.